File: app/src/main/java/cn/six/open/view/rv/OnRvItemClickListener.java
package cn.six.open.view.rv;

import android.support.v4.view.GestureDetectorCompat;
import android.support.v7.widget.RecyclerView;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;

/**
 * Created by songzhw on 2016-07-13
 *
 * NOTE: If you have a head wrapper, or a footer wrapper, or a load more wrapper,
 * please note that the position may different than usual!
 */
public abstract class OnRvItemClickListener implements RecyclerView.OnItemTouchListener {
    private GestureDetectorCompat gestureDetector;
    private RecyclerView rv;

    public OnRvItemClickListener(RecyclerView rv){
        this.rv = rv;
        gestureDetector = new GestureDetectorCompat(rv.getContext(), new RvGestureListener());
    }

    private class RvGestureListener extends GestureDetector.SimpleOnGestureListener{
        @Override
        public void onLongPress(MotionEvent e){
            // vibrator.vibrate(70); // Remember to add the Vibrator permission.

            View child = rv.findChildViewUnder(e.getX(), e.getY());
            if(child != null){
                RecyclerView.ViewHolder vh = rv.getChildViewHolder(child);
                onLongClick(vh);
            }
        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            View child = rv.findChildViewUnder(e.getX(), e.getY());
            if(child != null){
                RecyclerView.ViewHolder vh = rv.getChildViewHolder(child);
                onItemClick(vh);
            }
            return true;
        }
    }

    @Override
    public boolean onInterceptTouchEvent(RecyclerView rv, MotionEvent e) {
        gestureDetector.onTouchEvent(e);
        return false;
    }

    @Override
    public void onTouchEvent(RecyclerView rv, MotionEvent e) {
        gestureDetector.onTouchEvent(e);
    }

    @Override
    public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {
    }

    // ========================= abstract methods =================================
    public abstract void onItemClick(RecyclerView.ViewHolder vh);
    public abstract void onLongClick(RecyclerView.ViewHolder vh);
}

File: SixUiViews/lib/src/main/java/ca/six/views/rv/OnRvItemClickListener.kt
package ca.six.views.rv


import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import androidx.core.view.GestureDetectorCompat
import androidx.recyclerview.widget.RecyclerView


/**
 * Created by songzhw on 2016-07-13
 *
 * NOTE: If you have a head wrapper, or a footer wrapper, or a load more wrapper,
 * please note that the position may different than usual!
 */
abstract class OnRvItemClickListener(val rv: RecyclerView) : RecyclerView.OnItemTouchListener {
    private val gestureDetector: GestureDetectorCompat

    init {
        gestureDetector = GestureDetectorCompat(rv.context, RvGestureListener())
    }

    override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
        gestureDetector.onTouchEvent(e)
        return false
    }

    override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {
        gestureDetector.onTouchEvent(e)
    }

    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}

    fun onLongClick(vh: RecyclerView.ViewHolder?) {}

    private inner class RvGestureListener : SimpleOnGestureListener() {

        override fun onLongPress(e: MotionEvent) {
            val child = rv.findChildViewUnder(e.x, e.y)
            if (child != null) {
                val vh: RecyclerView.ViewHolder = rv.getChildViewHolder(child)
                onLongClick(vh)
            }
        }

        override fun onSingleTapUp(e: MotionEvent): Boolean {
            val child = rv.findChildViewUnder(e.x, e.y)
            if (child != null) {
                val vh: RecyclerView.ViewHolder = rv.getChildViewHolder(child)
                onItemClick(vh)
            }
            return true
        }
    }

    // ========================= abstract methods =================================
    abstract fun onItemClick(vh: RecyclerView.ViewHolder)


}